src/Symfony/Component/Messenger/Transport/Schedule/Progress/ProgressFactory.php

```
@@ -0,0 +1,79 @@
+<?php
+
+/*
+ * This file is part of the Symfony package.
+ *
+ * (c) Fabien Potencier <fabien@symfony.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Component\Messenger\Transport\Schedule\Progress;
+
+use Psr\Container\ContainerInterface;
+use Symfony\Component\Cache\Adapter\ArrayAdapter;
+use Symfony\Component\Clock\ClockInterface;
+use Symfony\Component\Lock\LockFactory;
+use Symfony\Component\Lock\LockInterface;
+use Symfony\Component\Lock\NoLock;
+use Symfony\Component\Messenger\Exception\LogicException;
+use Symfony\Contracts\Cache\CacheInterface;
+
+final class ProgressFactory implements ProgressFactoryInterface
+{
+    public function __construct(
+        private readonly ClockInterface $clock,
+        private readonly ContainerInterface $lockFactories,
+        private readonly ContainerInterface $caches,
+    ) {
+    }
+
+    public function create(string $scheduleName, array $options): ProgressInterface
+    {
+        $name = 'messenger.schedule.'.$scheduleName;
+        $lock = $this->createLock($scheduleName, $name, $options);
+        $cache = $this->createCache($scheduleName, $options);
+
+        return new Progress($this->clock, $lock, $cache, $name);
+    }
+
+    private function createLock(string $scheduleName, string $resourceName, array $options): LockInterface
+    {
+        if (!($options['lock'] ?? false)) {
```

If `symfony/lock` is not installed, we should throw an exception (as lock is not part of the required packages for Messenger).